infer/provider_builder_test.go

```diff
@@ -330,27 +335,21 @@ func TestValidate(t *testing.T) {
 //nolint:paralleltest // Running in parallel causes a data race.
 func TestBuildAndRun(t *testing.T) {
 	// 1. Create a provider without any components and ensure that it returns an error.
-	err := NewProviderBuilder().BuildAndRun()
+	_, err := NewProviderBuilder().Build()
 	require.Error(t, err)
 
-	// 2. Create a provider with a component and ensure that it starts successfully by starting the
-	// provider in a separate goroutine as it blocks the main thread.
-	errChan := make(chan error)
-
-	go func(errCh chan error) {
-		errCh <- NewProviderBuilder().
-			WithComponents(Component(NewMockComponentResource)).
-			WithName("test-provider").
-			WithVersion("1.0.0").
-			BuildAndRun()
-
-		close(errCh)
-	}(errChan)
-
-	select {
-	case err := <-errChan:
-		require.NoError(t, err, "provider startup should not fail")
-	case <-time.After(5 * time.Second):
-		return // The provider started successfully, so we can return.
-	}
+	// 2. Create a provider with a component and ensure that it starts and runs successfully.
+	p, err := NewProviderBuilder().
+		WithComponents(Component(NewMockComponentResource)).
+		WithName("test-provider").
+		WithVersion("1.0.0").
+		Build()
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	defer cancel()
+
+	err = p.Run(ctx)
+
+	assert.NoError(t, err, "provider startup should not fail")
```

Comment:
This is using https://github.com/pulumi/pulumi/pull/19445.

The assertion is stronger because we're now guaranteed the provider executed and returned a nil error.